apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
+          })
+        )
+    );
+  };
+}
 
-  /**
-   * Ejects out the front. Can eject from loaded or parked state. If there is
-   * no paper to eject, handler will do nothing and return positive acknowledgement.
-   */
-  async ejectPaper(): Promise<void> {
-    this.status = 'ejecting';
-    await this.driver.ejectPaper();
-  }
+function pollPaperStatus(): InvokeConfig<Context, PaperHandlerStatusEvent> {
+  return {
+    id: 'pollPaperStatus',
+    src: buildPaperStatusObservable(),
+  };
+}
 
-  /**
-   * Moves paper to print position and moves print head to DOWN position. If
-   * paper is already in an appropriate print position, does not move paper.
-   * E.g. if paper is loaded, it will pull the paper in a few more inches, but
-   * if the paper is parked, it will not move the paper. Printing can start in
-   * a variety of positions.
-   */
-  async enablePrint(): Promise<void> {
-    await this.driver.enablePrint();
-  }
+const NoPaperState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: pollPaperStatus(),
+  on: {
+    PAPER_INSIDE: {
+      // TODO make sure we don't print a ballot twice
+      target: 'waiting_for_ballot_data',
+    },
+    PAPER_READY_TO_LOAD: {
+      target: 'loading_paper',
+    },
+  },
+};
 
-  /**
-   * Moves print head to UP position, does not move paper
-   */
-  async disablePrint(): Promise<void> {
-    await this.driver.disablePrint();
-  }
+const LoadingPaperState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: pollPaperStatus(),
+  entry: (context) => {
+    // Paper can trigger sensors before it's actually able to be loaded,
+    // so we wait to decrease chance of failed load
+    setTimeout(async () => {
+      await context.driver.loadPaper();
+      await context.driver.parkPaper();
+    }, 300);
+  },
+  on: {
+    PAPER_INSIDE: {
+      target: 'waiting_for_ballot_data',
+    },
+  },
+};
 
-  async scanAndSave(pathOut: string): Promise<void> {
-    await this.driver.setScanDirection('backward');
-    const grayscaleResult = await this.driver.scan();
-    const grayscaleData = grayscaleResult.data;
-    const colorResult = createImageData(
-      grayscaleResult.width,
-      grayscaleResult.height
-    );
-    const colorData = new Uint32Array(
-      colorResult.data.buffer,
-      colorResult.data.byteOffset,
-      colorResult.data.byteLength
-    );
-    for (let i = 0; i < grayscaleData.byteLength; i += 1) {
-      const luminance = grayscaleData[i];
-      colorData[i] =
-        // eslint-disable-next-line no-bitwise
-        (luminance << 24) | (luminance << 16) | (luminance << 8) | 255;
-    }
-    await writeImageData(pathOut, colorResult);
-  }
+const WaitingForBallotDataState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: pollPaperStatus(),
```

Comment:
Do we need to poll for paper status if the only transition out of this state is for the voter to initiate a print?

Also, I'm not sure I understand what it means to wait for ballot data here. Could you give me a bit of context to help me understand?

Reply:
We don't need to poll paper status at the moment, and we're unlikely to in the next round of exception handling, so I'll remove this.

Waiting for ballot data means the state machine is waiting to receive the ballot image as a `Buffer` from the frontend. In terms of user flows, this means the blank paper has been loaded successfully and the voter is still making selections. The state machine is waiting for the user to finish their selections and press "Print", which sends the data and kicks off the print command.

Reply:
Thanks, that's helpful!